File: コンテスト/日鉄ソリューションズ プログラミングコンテスト 2022/C - Robot Takahashi/robot_takahashi.go
package main

import (
    "bufio"
    "fmt"
    "os"
)

func main()  {
    in := bufio.NewReader(os.Stdin)

    var n int
    fmt.Fscan(in, &n)

    var s string
    fmt.Fscan(in, &s)
    runes := []rune(s)

    weights := make([]int, n)
    for i := 0; i < n; i++ {
        fmt.Fscan(in, &weights[i])
    }

    best := 0
    max := 0
    for i := 0; i < n; i++ {
        threshold := weights[i]
        if threshold > max {
            max = threshold
        }

        curr := 0
        for j := 0; j < n; j++ {
            if weights[j] >= threshold && runes[j] == '1' {
                curr++
            } else if weights[j] < threshold && runes[j] == '0' {
                curr++
            }
        }

        if curr > best {
            best = curr
        }
    }

    curr := 0
    for j := 0; j < n; j++ {
        if runes[j] == '0' {
            curr++
        }
    }
    if curr > best {
        best = curr
    }

    fmt.Println(best)
}

File: コンテスト/日鉄ソリューションズ プログラミングコンテスト 2022/C - Robot Takahashi/robot_takahashi.py
#!/usr/bin/env python3

if __name__ == "__main__":
    n = int(input())
    s = input()
    weights = list(map(int, input().split()))

    pairs = [(weights[i], s[i]) for i in range(n)]
    pairs.sort()

    best = s.count("1")
    curr = best
    for i in range(n):
        if pairs[i][1] == "1":
            curr -= 1
        else:
            curr += 1
        if i < n-1:
            if pairs[i][0] != pairs[i+1][0]:
                best = max(curr, best)
        else:
            best = max(curr, best)
    best = max(best, s.count("0"))

    print(best)
